### DISCUSSION TITLE
deltaangle() method usage

### DISCUSSION THREAD

**Discussion by henrikas-svidras** [scikit-hep/vector#104] [Q&A] [answered]
Hello developers!

I wanted to thank you for developing this in Python. However, I was struggling to implement a feature while trying to follow your examples. I am not sure if this is the correct place to ask this, but let me know where this should be moved otherwise.

I have been recently trying out this package but I am completely puzzled by the `deltaangle(vector)` method. My goal was to get the 3D angle between two vectors. I naively expected the behaviour equivalent to that of `ROOT.TVector3.Angle()`, but the results were unclear to me. This is an example which illustrates my confusion:

```
>>> v1 = vector.obj(rho=0.6728646250485578,theta=1.669565,phi=2.25748)
>>> v1.mag
0.67616
>>> v2 = vector.obj(rho=3.765287519203256,theta=1.907928,phi=1.784361)
>>> v2.mag
3.989889
>>> v1.deltaangle(v2)
nan
```

The expected result would have been:
```
>>> root_v1 = ROOT.TVector3()
>>> root_v1.SetMagThetaPhi(0.67616,1.669565,2.25748)
>>> root_v2 = ROOT.TVector3()
>>> root_v2.SetMagThetaPhi(3.989889,1.907928,1.784361)
>>> root_v1.Angle(root_v2)
0.5184948169238227
```

Am I just misunderstanding something fundamental? Perhaps I initialise the vector object incorectly? Is it possible to get the same result as the one that I show in ROOT with this package as well? 

Thanks a lot.

**Comment by jpivarski** [comment, marked as answer]
Actually, you found a bug. (Delta angle is arccos of the dot product divided by both vectors' magnitudes, not just the dot product.)

There's an ongoing effort to test all of these methods against their ROOT equivalents in #36, but that isn't complete. It would have caught this error.

Meanwhile, PR #105 fixes it:

```python
>>> v1 = vector.obj(rho=0.6728646250485578,theta=1.669565,phi=2.25748)
>>> v2 = vector.obj(rho=3.765287519203256,theta=1.907928,phi=1.784361)
>>> v1.mag
0.67616
>>> v2.mag
3.989889
>>> v1.deltaangle(v2)
0.5184948169238227
```

**Reply by henrikas-svidras** [reply]
Thanks a lot for the prompt answer and the PR! I was not sure if this is a bug or misunderstanding of the method. But I am glad this was not too difficult to fix.

On a similar note, because this is related to my example, is there a way to initialise a 3D vector using the magnitude and angles, similar to SetMagThetaPhi() method from ROOT?

**Reply by jpivarski** [reply]
There isn't, though that would be nice. The way Vector's coordinate systems work, mag-phi-theta wouldn't be a possible storage type, but it could be a constructor. Right now, you can only construct vectors the way they are stored.

The reason it can't be a storage type is because the azimuthal coordinates can only depend on the vector's transverse projection; the longitudinal coordinates can only depend on the 3D projection, and the temporal coordinates can depend on everything. This is so that the planar code can be shared by all dimensions of vectors and the spatial code can be shared by 3D and 4D vectors. As another consequence of this, rapidity can't be a longitudinal coordinate, but it could be a (future) temporal coordinate.

This scheme allows for a large number of specializations—the above PR had to change 36 formulas for vectors in all combinations of allowed azimuthal and longitudinal coordinate systems (first vector can have 2 azimuthal × 3 longitudinal, so can the second vector, 6 × 6 = 36). ROOT's TVector3 only stores Cartesian coordinates and the ROOT::Math vectors define a handful of frequently used coordinate systems; we're trying to do something more general here.

Adding a constructor that immediately converts mag-phi-theta to rho-phi-theta wouldn't be any worse than immediately converting to Cartsian—it would lose some precision, but not as much. Phi and theta would be entirely unaffected. It would be better, though, to collect a full set of what constructors we'd like, so that they can be applied systematically across `vector.obj`, the NumPy backend, and the Awkward Array backend.

**Reply by henrikas-svidras** [reply]
Thanks a lot for the answers. I think it would be useful to have this. For now I will make a separate pT column myself, but if you can implement this sometime in the future, I would certainly take advantage of such a feature.

Thanks a lot again for your efforts, I'm glad I found this package, as it actually works much faster than a previous implementation with ROOT and pandas apply/lambda that I had.
